visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+> Note:  Accessors are also used to implement subscripts; the only real difference is that a subscript operation can have additional arguments.  This document will discuss property accesses and subscript operations interchangeably.
+
+## Two Dimensions of Access
+
+There are a wide variety of theoretically possible accessors, varying in what kind of access they provide and what mechanism they use to provide it.  Different programming languages support different combinations.  For Swift, there are three fundamental kinds of access that we are interested in:
+
+* Read access.
+* Write or “set” access.
+* Update access.
+
+Write and update access differ in that write access can create a value that did not exist before, while update access requires there be a pre-existing value.  For example, consider how the dictionary write operation `myDictionary["newKey"] = value` can create a value that did not exist before. In contrast, `myDictionary["existingKey"] += 1` modifies a value that must necessarily already exist.
+
+For each kind of access, there are multiple ways that the access might be provided.  The Swift implementation today supports three different possible mechanisms, although not all of these are currently supported by accessors:
+
+* Copying.  The value might be copied between the caller and the callee.
+* Borrowing.  The callee might provide access to the value without actually copying it.  This can be implemented by having the callee provide a pointer or reference to the value *in situ.*
+* Yielding.  Each of the above is typically provided by calling a function.  The value or pointer is passed as an argument or return value from that function.  It’s also possible for the callee to start a *coroutine* whose execution is interleaved with that of the caller.  This allows the callee to resume after the caller is finished, which is not possible for a function- or method-based accessor.
+
+## Six of Nine Fundamental Accessors
+
+Combining the possibilities above gives us a 3x3 matrix of possible accessors:
+
+|              |Read   |Write  |Update |
+|---           |---    |---    |---    |
+|**Copying**   |       |       |✗      |
+|**Borrowing** |       |✗      |       |
+|**Yielding**  |       |✗      |       |
+
+Three of these combinations turn out to be either impossible or not useful:
+
+* “Copying Update” is unnecessary.  A copying update would by definition copy the value into the caller and then copy the updated value back, which is easily achieved by using a “Copying Read” followed by a “Copying Write”
```

Comment:
Is that so categorically or only given present expressivity limits we place in the language? For example, if access can be async or throwing, is there some expressivity made possible with a unique "copying update" with respect to guaranteeing that there is/isn't an `async` operation between reading and writing, and/or that an update does/doesn't attempt to write if reading throws, etc.?

Reply:
As I hinted earlier, there are many other theoretically possible kinds of accessors that I am not describing here. My goal with this document was to find a (hopefully) clear way to explain how a certain set of six accessors seems to address Swift's current need to efficiently express the next generation of standard library APIs.  I would be interested in hearing more details about any use cases that you feel these do not cover.

For what it's worth:  The kind of guarantee you seem to be describing could be provided by a borrowing accessor plus a "swap" operator, or by a closure-taking update method.

Reply:
If you really think this grid is not helpful, I can delete this section.  For myself, this entire document started as an unstructured list of many different kinds of accessors and I struggled to try to explain how that large menagerie could be distilled down to a small set.  This grid was a key step in crystallizing my own thinking about this area:  Among other things, it helps explain why "set" is necessary regardless of what other accessors you have (it's the only option in the "write" column).    But different people approach these things differently, and a tool that was useful for me may not be useful for others.  I'll think about removing this section.